docs/site/Validation-controller-layer.md

```diff
@@ -0,0 +1,128 @@
+---
+lang: en
+title: 'Validation in the Controller Layer'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Validation-controller-layer.html
+---
+
+In the case where validation rules are not static, validations cannot be
+specified at the model level. Hence, validation can be added in the controller
+layer.
+
+Take an example of a promo code in an order, it is usually a defined value that
+only valid for a certain period of time. And in the CoffeeShop example, the area
+code of a phone number usually depends on the geolocation.
+
+## Add validation function in the Controller method
+
+The simplest way is to apply the validation function in the controller method,
+that is:
+
+```ts
+// create a validatePhoneNum function and call it here
+if (!this.validatePhoneNum(coffeeShop.phoneNum, coffeeShop.city))
+  throw new Error('Area code in phone number and city do not match.');
+return this.coffeeShopRepository.create(coffeeShop);
+```
+
+## Add interceptor for validation
+
+Another way is to use [interceptors](Interceptors.md).
+
+> Interceptors are reusable functions to provide aspect-oriented logic around
+> method invocations.
+
+Interceptors have access to the invocation context, including parameter values
+for the method call. It can perform more specific validations, for example,
+calling a service to check if an address is valid. There are three types of
+interceptors for different scopes: global, class-level and method-level
+interceptors.
+
+Interceptors can be created using the
+[interceptor generator](https://loopback.io/doc/en/lb4/Interceptor-generator.html)
+`lb4 interceptor` command. In the CoffeeShop example, the `phoneNum` in the
+`CoffeeShop` request body will be validated for the `POST` and `PUT` call
+whether the area code in the phone number matches the specified city. Since this
+is only applicable to the CoffeeShop endpoints, a non-global interceptor is
+created, i.e. specify `No` in the `Is it a global interceptor` prompt.
+
+```sh
+$ lb4 interceptor
+? Interceptor name: validatePhoneNum
+? Is it a global interceptor? No
+   create src/interceptors/validate-phone-num.interceptor.ts
+   update src/interceptors/index.ts
+
+Interceptor ValidatePhoneNum was created in src/interceptors/
+```
+
+In the newly created interceptor `ValidatePhoneNumInterceptor`, the function
+`intercept` is the place where the pre-invocation and post-invocation logic can
+be added.
+
+{% include code-caption.html content="/src/interceptors/validate-phone-num-interceptor.ts" %}
+
+```ts
+  async intercept(
+    invocationCtx: InvocationContext,
+    next: () => ValueOrPromise<InvocationResult>,
+  ) {
+    try {
+      // Add pre-invocation logic here
+      // ------ VALIDATE PHONE NUMBER ----------
+      const coffeeShop: CoffeeShop = new CoffeeShop();
```

Comment:
Why are we instantiating CoffeeShop here?  Won't all the fields be blank or zeroes?

Reply:
@dhmlau ^^

Reply:
`coffeeShop` will then be assigned to one of the invocation context argument depending on method call. 
Instead of calling:
```
const coffeeShop: CoffeeShop = new CoffeeShop();
```

I tried to use the following so that we don't need to instantiate:
```
let coffeeShop: CoffeeShop | undefined;
```
but got `500 TypeError: Cannot convert undefined or null to object`  error.  I guess it's ok to do it in the original code snippet?